### PR TITLE
create a workflow to smoke test package:flutter_markdown

### PR DESCRIPTION
**Pull request by devoncarew**
- create a workflow to smoke test `package:flutter_markdown`
- fix https://github.com/dart-lang/markdown/issues/477

~creating as a draft PR while I iron out issues~

### CODE REVIEW

**Comment by devoncarew** [comment]
Success!

It's failing, but that's expected (tracked in #480). Here's the issue in-lined:

```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following UnsupportedError was thrown building Directionality(textDirection: ltr):
Unsupported operation: Cannot add to an unmodifiable list

The relevant error-causing widget was:
  Directionality
  Directionality:file:///home/runner/work/markdown/markdown/flutter_packages/packages/flutter_markdown/test/utils.dart:164:10

When the exception was thrown, this was the stack:
#0      UnmodifiableListMixin.add (dart:_internal/list.dart:114:5)
#1      WikilinkSyntax.onMatch (file:///home/runner/work/markdown/markdown/flutter_packages/packages/flutter_markdown/test/custom_syntax_test.dart:165:18)
#2      InlineSyntax.tryMatch (package:markdown/src/inline_syntaxes/inline_syntax.dart:49:9)
#3      InlineParser.parse.<anonymous closure> (package:markdown/src/inline_parser.dart:121:43)
#4      ListMixin.any (dart:collection/list.dart:[149](https://github.com/dart-lang/markdown/actions/runs/3366377531/jobs/5582801560#step:9:150):15)
#5      InlineParser.parse (package:markdown/src/inline_parser.dart:121:20)
#6      Document.parseInline (package:markdown/src/document.dart:76:67)
#7      Document._parseInlineContent (package:markdown/src/document.dart:82:29)
#8      Document._parseInlineContent (package:markdown/src/document.dart:87:9)
#9      Document.parseLines (package:markdown/src/document.dart:71:5)
#10     _MarkdownWidgetState._parseMarkdown (package:flutter_markdown/src/widget.dart:320:45)
#11     _MarkdownWidgetState.didChangeDependencies (package:flutter_markdown/src/widget.dart:284:5)
#12     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5107:11)
#13     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4932:5)
...     Normal element mounting (7 frames)
```

**Comment by devoncarew** [comment]
I believe the actual https://github.com/flutter/packages repo may be tested against flutter stable and flutter master? We're testing against flutter beta; the skew may become relevant at some point.

**Comment by devoncarew** [comment]
> @devoncarew – so yay it's failing. Now what?

Sam has a fix upstream - https://github.com/flutter/packages/pull/2758. We can land this after.

File: .github/workflows/flutter_markdown.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+          path: markdown
+
+      - name: clone flutter/packages
+        uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b
+        with:
+          repository: flutter/packages
+          path: flutter_packages
+
+      - name: get flutter
+        uses: subosito/flutter-action@dbf1fa04f4d2e52c33185153d06cdb5443aa189d
```
Comment (srawlins) [inline .github/workflows/flutter_markdown.yaml:34]: Can you say what this is?

**Review by srawlins — APPROVED** [review]
Very cool!

File: .github/workflows/flutter_markdown.yaml (hunk 1 above)
Comment (kevmoo) [reply .github/workflows/flutter_markdown.yaml:34]: Installs the Flutter SDK?

Comment (srawlins) [reply .github/workflows/flutter_markdown.yaml:34]: Haha I get that. I just don't know what "subosito" is .

Comment (devoncarew) [reply .github/workflows/flutter_markdown.yaml:34]: It seems to be the standard flutter 'install sdk' action; I'll add a comment here however.

**Comment by devoncarew** [comment]
I'm seeing:

```
table_test.dart: Table should work with alignments                                                  
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: TextAlign:<TextAlign.right>
  Actual: TextAlign:<TextAlign.left>

When the exception was thrown, this was the stack:
#4      defineTests.<anonymous closure>.<anonymous closure> (file:///home/runner/work/markdown/markdown/flutter_packages/packages/flutter_markdown/test/table_test.dart:61:9)
```

and potentially two copies of:

```
flutter_markdown/test/table_test.dart: Table GFM Examples remainder of table cells may vary, excess cells are ignored  
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following UnsupportedError was thrown building Directionality(textDirection: ltr):
Unsupported operation: Cannot add to an unmodifiable list

The relevant error-causing widget was:
  Directionality
  Directionality:file:///home/runner/work/markdown/markdown/flutter_packages/packages/flutter_markdown/test/utils.dart:164:10

When the exception was thrown, this was the stack:
#0      UnmodifiableListMixin.add (dart:_internal/list.dart:114:5)
#1      MarkdownBuilder.visitElementBefore (package:flutter_markdown/src/builder.dart:263:27)
...
```

**Comment by srawlins** [comment]
Perhaps there are remaining new incompatibilities with the two packages.

**Comment by devoncarew** [comment]
> Perhaps there are remaining new incompatibilities with the two packages.

It looks like that's the case. I could adjust this new job to be `continue-on-error: true`, and track the new failures in an issue (we'd revert the allow failure once the new issues are addressed)?

**Comment by devoncarew** [comment]
Hmm, it looks like `continue-on-error: true` doesn't do what I was hoping it would do (have the bot pass even on failures).

**Comment by kevmoo** [comment]
@devoncarew – I'd say go ahead and land the `anyOf` thing we chatted about in `flutter_markdown`. Validate locally, etc.

**Comment by devoncarew** [comment]
@chenzhiguang - some of the issues _may_ have been introduced by https://github.com/dart-lang/markdown/pull/472. I don't know if the parts of that PR should be reverted, or, if we should address the issues downstream in package:flutter_markdown itself (if it's relying on either implementation details or previous bugs from markdown).

In either case, it would be nice to get the repos to a point where we can land this PR and then incrementally work from there.

**Comment by kevmoo** [comment]
@devoncarew I did a bisect. It's https://github.com/dart-lang/markdown/pull/472 that started the breaks.

**Comment by chenzhiguang** [comment]
@devoncarew @kevmoo I am working on it know.

**Comment by kevmoo** [comment]
@devoncarew – well, it looks like we moved from `style="text-align: xyz` to `align=xyz` per spec

This seems to be a big reason for the failure

**Comment by kevmoo** [comment]
```diff
@@ -470,11 +472,11 @@ class MarkdownBuilder implements md.NodeVisitor {
         current.children.add(_buildRichText(const TextSpan(text: '\n')));
       } else if (tag == 'th' || tag == 'td') {
         TextAlign? align;
-        final String? style = element.attributes['style'];
+        final String? style = element.attributes['align'];
         if (style == null) {
           align = tag == 'th' ? styleSheet.tableHeadAlign : TextAlign.left;
         } else {
-          final RegExp regExp = RegExp(r'text-align: (left|center|right)');
+          final RegExp regExp = RegExp(r'(left|center|right)');
           final Match match = regExp.matchAsPrefix(style)!;
           switch (match[1]) {
             case 'left':
```

this fixes one of the bugs

The other thing: `flutter_markdown` MUTATES the element tree, but @chenzhiguang makes `children` in one case `const` which is another failure...

**Comment by srawlins** [comment]
Thanks for catching that Kevin; I missed this in the review; we _should_ be sticking to `<td style="text-align"` rather than `td align="`. We should revert that part of the PR. I've commented on it.

**Comment by chenzhiguang** [comment]
> ```diff
> +          final RegExp regExp = RegExp(r'(left|center|right)');
>            final Match match = regExp.matchAsPrefix(style)!;
>            switch (match[1]) {
> ```
Yep, I found it too. I think here  `switch (style)` is enought.

**Comment by kevmoo** [comment]
> Thanks for catching that Kevin; I missed this in the review; we _should_ be sticking to `<td style="text-align"` rather than `td align="`. We should revert that part of the PR. I've commented on it.

See my comment on the original PR

> "deprecated" is tricky. This syntax will be supported FOR EVER. If there are zero concerns about it working, sticking to spec will save us a lot of sanity in the future IMHO

**Comment by chenzhiguang** [comment]
Another failure is caused by this change too:
```diff
- children.add(Element.empty('td'));
+ children.add(Element('td', const []));
```
`Element.empty('td')` has a `Null` `children`, while `children.add(Element('td', const []))` has an empty `children`.

**Comment by chenzhiguang** [comment]
Do I need to do any change on our side?

**Comment by kevmoo** [comment]
> Do I need to do any change on our side?

@chenzhiguang – I pushed one small change on our side (dropped `const []` in favor of `[]`) – other fixes should be in flutter_markdown, I think!

**Comment by kevmoo** [comment]
@devoncarew @srawlins @chenzhiguang - https://github.com/flutter/packages/pull/2777

**Comment by kevmoo** [comment]
Haza! Fixed @devoncarew !!

File: lib/src/block_syntaxes/table_syntax.dart [hunk 2]
```diff
@@ -51,7 +51,7 @@ class TableSyntax extends BlockSyntax {
       if (children != null) {
         while (children.length < columnCount) {
           // Insert synthetic empty cells.
-          children.add(Element('td', const []));
```
Comment (kevmoo) [inline lib/src/block_syntaxes/table_syntax.dart:54]: Smallest diff to fix `flutter_markdown` integration

**Pull request merged**